### DISCUSSION TITLE
How to use corfu-history.el which is located in a subdirectory of the corfu-repository?

### DISCUSSION THREAD

**Discussion by mekeor** [raxod502/straight.el#968] [Q&A] [answered]
I installed [corfu](https://github.com/minad/corfu) using use-package and straight.el. In its [repository](https://github.com/minad/corfu), corfu has some elisp-files located in a [subdirectory called `extensions`](https://github.com/minad/corfu/tree/main/extensions), e.g. [`extensions/corfu-history.el`](https://github.com/minad/corfu/blob/main/extensions/corfu-history.el). But after the installation, that subdirectory is not added to the load-path and further, corfu-history.el is not loaded.

So, how can I use corfu-history.el without having to manually load the file?

In contrast, on another machine, I installed straight.el using GNU Guix which added both corfu.el and all extension-elisp-files into a single folder which was recursively loaded and added to load-path.

**Comment by progfolio** [comment, marked as answer]
Have you added the extensions directory to the recipe's `:files`?  e.g.

```emacs-lisp
(corfu :files (:defaults "extensions/*"))
```

**Reply by mekeor** [reply]
I used `M-x straight-get-recipe corfu RET` to get the default recipe. I added `"extensions/*.el` to the `:files` section, as you described. Finally, I had `:straight (corfu :type git :host github :repo "emacs-straight/corfu" :files ("*" "extensions/*.el" (:exclude ".git")))`. It worked, thank you!

**Comment by publicimageltd** [comment]
I am having the same problem, and I do not feel that this is a good answer. What if the recipe stored online changes? We won't notice it since we have our own recipe now.

Is there an automatic way to modify the recipe so that it only *adds* the extensions to the loadpath, without having to manually set the whole recipe?

**Reply by progfolio** [reply]
```emacs-lisp
(corfu :files (:defaults "extensions"))
```
Should cover this particular case without copying the rest of the recipe. The rest of the recipe will be inherited.

> What if the recipe stored online changes? We won't notice it since we have our own recipe now.

Limiting what you override should cover most cases. Updating recipe repositories will update the inherited recipe.

**Reply by publicimageltd** [reply]
Thanks, that's exactly what I was looking for!

**Reply by publicimageltd** [reply]
Sorry, this is actually a straight question, but yet: It doesn't work. If I use the shorter declaration, the extensions directory is not added to my load-path. Yet it is built correctly (I looked in to the build directory). What is weird, also, is that I can't get the recipe for corfu using `straight-get-recipe`. It's not offered as a choice. And still I can use corfu, it's there and in the load-path (w/o the extensions, that is).

If I find time, I'll repost this problem in a more orderly fashion and with `emacs -Q` minimal example as an issue for straight, since this is not the right place. I am hoping, however, that someone reading this might have an idea of what my problem could be.

**Comment by progfolio** [comment]
> If I use the shorter declaration, the extensions directory is not added to my load-path.

It shouldn't need to be on your load-path for this to work.
The files directive should copy the relevant files to straight/build/corfu (which will be on the load-path).
Probably just needs to be adjusted to:  `:files (:defaults "extensions/*.el")` 

> What is weird, also, is that I can't get the recipe for corfu using straight-get-recipe. It's not offered as a choice.

Sounds like you are on the "master" branch. This is fixed on the "develop" branch. You can switch by adding `(setq straight-repository-branch "develop")` prior to the bootstrapping snippet and rebuilding straight.el.

**Reply by publicimageltd** [reply]
The `*.el` did the trick! Thanks a lot!
